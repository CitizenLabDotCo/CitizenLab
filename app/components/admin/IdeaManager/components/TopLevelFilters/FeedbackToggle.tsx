import React from 'react';
import { adopt } from 'react-adopt';
import { isNilOrError } from 'utils/helperUtils';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from '../../messages';

// styles
import styled, { css } from 'styled-components';
import { colors, fontSizes } from 'utils/styleUtils';

// resources
import GetIdeasCount, { GetIdeasCountChildProps } from 'resources/GetIdeasCount';

// components
import CountBadge from 'components/UI/CountBadge';

const size = 21;
const padding = 4;

const Container = styled.div`
  display: inline-block;
  display: flex;
  align-items: center;
`;

const ToggleContainer: any = styled.div`
  height: 100%;
  display: flex;
  align-items: center;

  ${(props: any) => props.disabled && css`
    opacity: 0.25;

    i,
    i:before {
      cursor: not-allowed;
    }
  `};

  ${(props: any) => props.checked && css`
    i {
      padding-right: ${padding}px !important;
      padding-left: ${size}px !important;
      background: ${colors.clGreen} !important;
    }
  `};

  input {
    display: none;
  }

  i {
    display: inline-block;
    cursor: pointer;
    padding: ${padding}px;
    padding-right: ${size}px;
    transition: all ease 0.15s;
    border-radius: ${size + padding}px;
    background: #ccc;
    transform: translate3d(0, 0, 0);

    &:before {
      display: block;
      content: '';
      width: ${size}px;
      height: ${size}px;
      border-radius: ${size}px;
      background: #fff;
    }
  }
`;

const StyledLabel = styled.label`
  font-size: ${fontSizes.small}px;
  font-weight: 400;
  line-height: 20px;
  padding-left: 10px;
  padding-right: 10px;
  cursor: pointer;
  display: flex;
  align-items: center;
`;

interface InputProps {
  value: boolean;
  onChange: () => void;
  assignee: string;
}

interface DataProps {
  feedbackNeededCount: GetIdeasCountChildProps;
}

interface Props extends InputProps, DataProps {}

type State = {};

export class FeedbackToggle extends React.PureComponent<Props, State> {
  handleOnClick = () => {
    this.props.onChange();
  }

  render() {
    const { value, feedbackNeededCount } = this.props;

    return (
      <Container className="feedback_needed_filter_toggle">
        <StyledLabel onClick={this.handleOnClick}>
          <FormattedMessage {...messages.anyFeedbackStatus} />
        </StyledLabel>
        <ToggleContainer onClick={this.handleOnClick} checked={value}>
          <input type="checkbox" role="checkbox" aria-checked={value}/>
          <i />
        </ToggleContainer>
        <StyledLabel onClick={this.handleOnClick}>
          <FormattedMessage {...messages.needFeedback} />
          {!isNilOrError(feedbackNeededCount.count) && <CountBadge count={feedbackNeededCount.count}/>}
        </StyledLabel>
      </Container>
    );
  }
}

const Data = adopt({
  ideasCount: ({ assignee, render }) => (
    <GetIdeasCount feedbackNeeded={true} assignee={assignee === 'all' ? undefined : assignee}>
      {render}
    </GetIdeasCount>
  )
});

export default (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <FeedbackToggle {...inputProps} feedbackNeededCount={dataProps.ideasCount}/>}
  </Data>
);
