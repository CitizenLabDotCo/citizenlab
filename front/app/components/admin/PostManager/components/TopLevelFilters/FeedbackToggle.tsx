import React from 'react';
import { adopt } from 'react-adopt';
import { isNilOrError } from 'utils/helperUtils';
import { isFunction } from 'lodash-es';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from '../../messages';

// styles
import styled, { css } from 'styled-components';
import { colors, fontSizes } from 'utils/styleUtils';

// resources
import GetIdeasCount, {
  GetIdeasCountChildProps,
} from 'resources/GetIdeasCount';
import GetInitiativesCount, {
  GetInitiativesCountChildProps,
} from 'resources/GetInitiativesCount';

// typings
import { ManagerType } from '../..';

// components
import CountBadge from 'components/UI/CountBadge';

const size = 21;
const padding = 4;

const Container = styled.div`
  display: inline-block;
  display: flex;
  align-items: center;
`;

const ToggleContainer: any = styled.div`
  height: 100%;
  display: flex;
  align-items: center;

  ${(props: any) =>
    props.disabled &&
    css`
      opacity: 0.25;

      i,
      i:before {
        cursor: not-allowed;
      }
    `};

  ${(props: any) =>
    props.checked &&
    css`
      i {
        padding-right: ${padding}px !important;
        padding-left: ${size}px !important;
        background: ${colors.clGreen} !important;
      }
    `};

  input {
    display: none;
  }

  i {
    display: inline-block;
    cursor: pointer;
    padding: ${padding}px;
    padding-right: ${size}px;
    transition: all ease 0.15s;
    border-radius: ${size + padding}px;
    background: #ccc;
    transform: translate3d(0, 0, 0);

    &:before {
      display: block;
      content: '';
      width: ${size}px;
      height: ${size}px;
      border-radius: ${size}px;
      background: #fff;
    }
  }
`;

const StyledLabel = styled.label`
  font-size: ${fontSizes.small}px;
  font-weight: 400;
  line-height: 20px;
  padding-left: 10px;
  padding-right: 10px;
  cursor: pointer;
  display: flex;
  align-items: center;
`;

interface InputProps {
  type: ManagerType;
  value: boolean;
  onChange: (feedbackNeeded: boolean | undefined) => void;
  assignee?: string | null;
  project?: string | null;
  phase?: string | null;
  topics?: string[] | null;
  status?: string | null;
  searchTerm?: string | null;
}

interface DataProps {
  feedbackNeededCount: GetIdeasCountChildProps | GetInitiativesCountChildProps;
}

interface Props extends InputProps, DataProps {}

interface State {}

export class FeedbackToggle extends React.PureComponent<Props, State> {
  componentDidUpdate(prevProps) {
    if (prevProps.searchTerm !== this.props.searchTerm) {
      if (isFunction(this.props.feedbackNeededCount.onChangeSearchTerm)) {
        this.props.feedbackNeededCount.onChangeSearchTerm(
          this.props.searchTerm || ''
        );
      }
    }
  }

  handleOnClick = () => {
    this.props.onChange(!this.props.value || undefined);
  };

  render() {
    const { value, feedbackNeededCount } = this.props;

    return (
      <Container
        id="e2e-feedback_needed_filter_toggle"
        className="feedback_needed_filter_toggle"
      >
        <ToggleContainer onClick={this.handleOnClick} checked={value}>
          <input type="checkbox" role="checkbox" aria-checked={value} />
          <i />
        </ToggleContainer>
        <StyledLabel onClick={this.handleOnClick}>
          <FormattedMessage {...messages.inputsNeedFeedbackToggle} />
          {!isNilOrError(feedbackNeededCount.count) && (
            <CountBadge count={feedbackNeededCount.count} />
          )}
        </StyledLabel>
      </Container>
    );
  }
}

const Data = adopt({
  feedbackNeededCount: ({
    project,
    phase,
    topics,
    status,
    assignee,
    render,
    type,
  }) => {
    const projectIds = [project];

    return type === 'Initiatives' ? (
      <GetInitiativesCount
        feedbackNeeded={true}
        assignee={assignee}
        topics={topics}
        initiativeStatusId={status}
      >
        {render}
      </GetInitiativesCount>
    ) : (
      <GetIdeasCount
        feedbackNeeded={true}
        assignee={assignee}
        projectIds={projectIds}
        phaseId={phase}
        topics={topics}
        ideaStatusId={status}
      >
        {render}
      </GetIdeasCount>
    );
  },
});

export default (inputProps: InputProps) => (
  <Data {...inputProps}>
    {(dataProps) => <FeedbackToggle {...inputProps} {...dataProps} />}
  </Data>
);
