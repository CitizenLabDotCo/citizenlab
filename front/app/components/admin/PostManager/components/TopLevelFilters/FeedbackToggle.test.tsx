// @ts-nocheck
import React from 'react';
import { shallow } from 'enzyme';

import { FeedbackToggle } from './FeedbackToggle';

jest.mock('services/stats');
jest.mock('utils/cl-intl');

// mocking dependencies
jest.mock('resources/GetIdeasCount', () => 'GetIdeasCount');
jest.mock('resources/GetInitiativesCount', () => 'GetInitiativesCount');
jest.mock('components/UI/CountBadge', () => 'CountBadge');
jest.mock('utils/cl-intl');
jest.mock('modules', () => ({ streamsToReset: [] }));

import 'jest-styled-components';

describe('<FeedbackToggle />', () => {
  let onChange: jest.Mock;
  beforeEach(() => {
    onChange = jest.fn();
  });

  it('renders correctly unchecked for ideas', () => {
    const ideasCount = { count: 6 };

    const wrapper = shallow(
      <FeedbackToggle
        type="AllIdeas"
        assignee="me"
        feedbackNeededCount={ideasCount}
        value={false}
        onChange={onChange}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly checked for initiatives', () => {
    const initiativesCount = { count: 6 };

    const wrapper = shallow(
      <FeedbackToggle
        type="Initiatives"
        assignee="me"
        feedbackNeededCount={initiativesCount}
        value={true}
        onChange={onChange}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('has coherent aria checked attibutes', () => {
    const ideasCount = { count: 6 };

    const wrapper = shallow(
      <FeedbackToggle
        type="AllIdeas"
        assignee="me"
        feedbackNeededCount={ideasCount}
        value={true}
        onChange={onChange}
      />
    );
    const Toggle = wrapper.find('FeedbackToggle__ToggleContainer');
    expect(Toggle.props().checked).toBeTruthy;
    const Input = wrapper.find('input');
    expect(Input.props()['aria-checked']).toBeTruthy;
  });

  it('calls onChange when clicked', () => {
    const ideasCount = { count: 6 };

    const wrapper = shallow(
      <FeedbackToggle
        type="Initiatives"
        assignee="me"
        feedbackNeededCount={ideasCount}
        value={false}
        onChange={onChange}
      />
    );
    wrapper
      .find('FeedbackToggle__ToggleContainer')
      .simulate('click', { preventDefault() {}, stopPropagation() {} });

    expect(onChange).toHaveBeenCalled();
  });

  it('reacts to search change', () => {
    const onChangeSearchTerm = jest.fn();
    const ideasCount = {
      onChangeSearchTerm,
      count: 6,
    };

    const wrapper = shallow(
      <FeedbackToggle
        assignee="me"
        feedbackNeededCount={ideasCount}
        value={false}
        onChange={onChange}
      />
    );
    wrapper.setProps({ searchTerm: 'lalalalalala' });

    expect(onChangeSearchTerm).toHaveBeenCalledWith('lalalalalala');
  });
});
