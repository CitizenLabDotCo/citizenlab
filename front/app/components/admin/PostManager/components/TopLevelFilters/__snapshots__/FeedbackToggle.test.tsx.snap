// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedbackToggle /> renders correctly checked for initiatives 1`] = `
<FeedbackToggle__Container
  className="feedback_needed_filter_toggle"
  id="e2e-feedback_needed_filter_toggle"
>
  <FeedbackToggle__ToggleContainer
    checked={true}
    onClick={[Function]}
  >
    <input
      aria-checked={true}
      role="checkbox"
      type="checkbox"
    />
    <i />
  </FeedbackToggle__ToggleContainer>
  <FeedbackToggle__StyledLabel
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Only show inputs that need feedback"
      id="app.components.admin.PostManager.inputsNeedFeedbackToggle"
    />
    <CountBadge
      count={6}
    />
  </FeedbackToggle__StyledLabel>
</FeedbackToggle__Container>
`;

exports[`<FeedbackToggle /> renders correctly unchecked for ideas 1`] = `
<FeedbackToggle__Container
  className="feedback_needed_filter_toggle"
  id="e2e-feedback_needed_filter_toggle"
>
  <FeedbackToggle__ToggleContainer
    checked={false}
    onClick={[Function]}
  >
    <input
      aria-checked={false}
      role="checkbox"
      type="checkbox"
    />
    <i />
  </FeedbackToggle__ToggleContainer>
  <FeedbackToggle__StyledLabel
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Only show inputs that need feedback"
      id="app.components.admin.PostManager.inputsNeedFeedbackToggle"
    />
    <CountBadge
      count={6}
    />
  </FeedbackToggle__StyledLabel>
</FeedbackToggle__Container>
`;
